design/cryptography.md

```diff
@@ -0,0 +1,127 @@
+- Transactions using AccessList can specify the desired revelation period
+- Using Null Address combined with hexadecimal 1-5 level determines the revelation period
+- Obscuro interprets the revelation period and encrypts the transaction with the corresponding revelation period key
+
+The [specification](https://eips.ethereum.org/EIPS/eip-2930) of the `AccessList` Field is as follows:
+
+```
+For the transaction to be valid, accessList must be of type [[{20 bytes}, [{32 bytes}...]]...], 
+where ... means “zero or more of the thing to the left”.
+For example, the following is a valid access list (all hex strings would in reality be in byte representation):
+
+[
+    [
+        "0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae",
+        [
+            "0x0000000000000000000000000000000000000000000000000000000000000003",
+            "0x0000000000000000000000000000000000000000000000000000000000000007"
+        ]
+    ],
+    [
+        "0xbb9bc244d798123fde783fcc1c72d3bb8c189413",
+        []
+    ]
+]
+```
+
+Obscuro uses the `AccessList` object to determine the revelation period.
+A special address and special storage keys combination is checked for internally.
+
+To simplify the protocol the Null address (0x0000000000000000000000000000000000000000) will be used and the hexadecimal representation of 1 to 5 is used to specify the revelation period.
+
+An example would be:
+```
+[
+    [
+        "0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae",                             <- Canonical AccessList usage
+        [
+            "0x0000000000000000000000000000000000000000000000000000000000000003",
+            "0x0000000000000000000000000000000000000000000000000000000000000007"
+        ]
+    ],
+    [
+        "0x0000000000000000000000000000000000000000",                             <- Obscuro Revelation Period
+         [
+            "0x0000000000000000000000000000000000000000000000000000000000000003", <- 1 day revelation Period
+            "0x0000000000000000000000000000000000000000000000000000000000000007"  <- ignored
+        ]
+    ]
+]
+```
+
+
+### 4. Reveal the rollup transactions
+- L1 blocks are used as a clock mechanism
+  - There are 5 revelation periods
+    * XS - 12 seconds (1 block)
+    * S - 1 hour (3600/12 = 300 blocks)
+    * M - 1 day (24 * 3600/12 = 7200 blocks)
+    * L - 1 month (30 * 24 * 3600/12 = 216000 blocks)
+    * XL - 1 year (12 * 30 * 24 * 3600/12 = 2592000 blocks)
+- Validators do not reveal keys so that keys are not susceptible to attacks
+- Central Sequencer stores the keys in a rollup, decrypt time, key tuple in a database.
+- When a Light Batch is created the revealed keys are appended to it and removed from the database.
+- If the database is lost for some reason, the Central Sequencer can recalculate the keys on-demand.
+
+
+### Problems
+
+1. Symmetric vs Asymmetric encryption for rollup tx encryption.
+- symmetric is more space efficient
+- AES 256 is the battle tested standard
```

Comment:
why not a 128bit key?

Reply:
No particular reason as I didn't dig into the key size. I think that should be done in the testing/performance phase ?

Do you think we should use 128bit key ?